fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneHighlighting.java

```diff
@@ -50,127 +50,214 @@
-                                pres.poll();
-                            }
-                            if (ends.size() < luceneQueryHighlightParameters.getSnippedSize() - tokenCountAfterHighlighted) {
-                                ends.add(text.substring(upto, startOffset));
-                            }
-                        }
-                    } else {
-                        addNonMatch(sb, text.substring(upto, startOffset));
-                    }
-                    upto = startOffset;
-                } else if (upto > startOffset) {
-                    continue;
-                }
+        }
 
-                if (matchedTokens.contains(token)) {
-                    // Token matches.
-                    if (luceneQueryHighlightParameters.isCutSnippets() && currentPos - lastMatchPos > tokenCountBeforeHighlighted + tokenCountAfterHighlighted + 1) {
-                        addNonMatch(sb, highlightedTextConnector);
-                    }
-                    while (!pres.isEmpty()) {
-                        addNonMatch(sb, pres.poll());
-                    }
-                    ends.clear();
-                    int start = startOffset;
-                    while (start < endOffset) {
-                        int index = text.toLowerCase(Locale.ROOT).indexOf(token, start);
-                        if (index < 0 || index >= endOffset) {
-                            addNonMatch(sb, text.substring(start, endOffset));
-                            break;
-                        }
-                        int actualEndOffset = index + token.length();
-                        addNonMatch(sb, text.substring(start, index));
-                        String substring = text.substring(index, actualEndOffset);
-                        if (substring.equalsIgnoreCase(token)) {
-                            addWholeMatch(sb, substring,
-                                    highlightedPositions);
-                        } else {
-                            addNonMatch(sb, substring);
-                        }
-                        start = actualEndOffset;
-                    }
-                    upto = endOffset;
-                    matchedInText.add(token);
-                    lastMatchPos = currentPos;
-                } else {
-                    for (String prefixToken : prefixTokens) {
-                        if (token.startsWith(prefixToken)) {
-                            addPrefixMatch(sb, text.substring(startOffset, endOffset), prefixToken,
-                                    highlightedPositions);
-                            upto = endOffset;
-                            matchedPrefixes.add(prefixToken);
-                            break;
-                        }
-                    }
+        boolean next() throws IOException {
+            int oldStartOffset = startOffset();
+            int oldEndOffset = endOffset();
+            while(true) {
```

Comment:
Could you please comment this will not loop forever in case a synonym "chain" exists is at the end, e.g. something like: `the colour is purple {violet, mauve}`.